crypto/verification_sas_methods.go

```diff
@@ -0,0 +1,143 @@
+	{'🐶', "Dog"},
+	{'🐱', "Cat"},
+	{'🦁', "Lion"},
+	{'🐎', "Horse"},
+	{'🦄', "Unicorn"},
+	{'🐷', "Pig"},
+	{'🐘', "Elephant"},
+	{'🐰', "Rabbit"},
+	{'🐼', "Panda"},
+	{'🐓', "Rooster"},
+	{'🐧', "Penguin"},
+	{'🐢', "Turtle"},
+	{'🐟', "Fish"},
+	{'🐙', "Octopus"},
+	{'🦋', "Butterfly"},
+	{'🌷', "Flower"},
+	{'🌳', "Tree"},
+	{'🌵', "Cactus"},
+	{'🍄', "Mushroom"},
+	{'🌏', "Globe"},
+	{'🌙', "Moon"},
+	{'☁', "Cloud"},
+	{'🔥', "Fire"},
+	{'🍌', "Banana"},
+	{'🍎', "Apple"},
+	{'🍓', "Strawberry"},
+	{'🌽', "Corn"},
+	{'🍕', "Pizza"},
+	{'🎂', "Cake"},
+	{'❤', "Heart"},
+	{'😀', "Smiley"},
+	{'🤖', "Robot"},
+	{'🎩', "Hat"},
+	{'👓', "Glasses"},
+	{'🔧', "Spanner"},
+	{'🎅', "Santa"},
+	{'👍', "Thumbs Up"},
+	{'☂', "Umbrella"},
+	{'⌛', "Hourglass"},
+	{'⏰', "Clock"},
+	{'🎁', "Gift"},
+	{'💡', "Light Bulb"},
+	{'📕', "Book"},
+	{'✏', "Pencil"},
+	{'📎', "Paperclip"},
+	{'✂', "Scissors"},
+	{'🔒', "Lock"},
+	{'🔑', "Key"},
+	{'🔨', "Hammer"},
+	{'☎', "Telephone"},
+	{'🏁', "Flag"},
+	{'🚂', "Train"},
+	{'🚲', "Bicycle"},
+	{'✈', "Aeroplane"},
+	{'🚀', "Rocket"},
+	{'🏆', "Trophy"},
+	{'⚽', "Ball"},
+	{'🎸', "Guitar"},
+	{'🎺', "Trumpet"},
+	{'🔔', "Bell"},
+	{'⚓', "Anchor"},
+	{'🎧', "Headphones"},
+	{'📁', "Folder"},
+	{'📌', "Pin"},
+}
+
+// VerificationEmoji describes an emoji that might be sent for verifying devices.
+type VerificationEmoji struct {
+	Emoji       rune
+	Description string
+}
+
+// GetSASVerificationNumbers generates the three numbers that need to match with the other device for a verification to be valid.
+func (mach *OlmMachine) GetSASVerificationNumbers(initUserID id.UserID, initDeviceID id.DeviceID,
+	acceptUserID id.UserID, acceptDeviceID id.DeviceID, transactionID string, sas *olm.SAS) ([3]uint, error) {
+	sasInfo := "MATRIX_KEY_VERIFICATION_SAS" +
+		initUserID.String() + initDeviceID.String() +
+		acceptUserID.String() + acceptDeviceID.String() +
+		transactionID
+
+	sasBytes, err := sas.GenerateBytes([]byte(sasInfo), 5)
+	if err != nil {
+		return [3]uint{0, 0, 0}, err
+	}
+
+	numbers := [3]uint{
+		(uint(sasBytes[0])<<5 | uint(sasBytes[1])>>3) + 1000,
+		(uint(sasBytes[1]&0x7)<<10 | uint(sasBytes[2])<<2 | uint(sasBytes[3]>>6)) + 1000,
+		(uint(sasBytes[3]&0x3F)<<7 | uint(sasBytes[4])>>1) + 1000,
+	}
+	mach.Log.Debug("Generated SAS numbers are: %v-%v-%v", numbers[0], numbers[1], numbers[2])
+
+	return numbers, nil
+}
+
+// GetSASVerificationEmojis generates the three numbers that need to match with the other device for a verification to be valid.
+func (mach *OlmMachine) GetSASVerificationEmojis(initUserID id.UserID, initDeviceID id.DeviceID,
+	acceptUserID id.UserID, acceptDeviceID id.DeviceID, transactionID string, sas *olm.SAS) ([7]VerificationEmoji, error) {
+
+	sasInfo := "MATRIX_KEY_VERIFICATION_SAS" +
+		initUserID.String() + initDeviceID.String() +
+		acceptUserID.String() + acceptDeviceID.String() +
+		transactionID
+
+	var emojis [7]VerificationEmoji
+	sasBytes, err := sas.GenerateBytes([]byte(sasInfo), 6)
+	emojisStr := ""
```

Comment:
Is `emojisStr` used for anything else than debugging? If not, you could probably just log the `emojis` array instead of building a string manually

Reply:
Indeed, removed